File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.cpp
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        return solve(points);
    }
private:
    // Daily
    int solve(vector<vector<int>>& points) {
        int n = points.size();
        int i = 0, ans = 0, cp = 0; // cp- connected points
        vector<bool> visited(n, false);
        priority_queue<pair<int, int>> minHeap;
        
        while(++cp < n) {
            visited[i] = true;
            for(int j = 0; j < n; ++j) {
                if(!visited[j]) {
                    int dist = abs(points[i][1] - points[j][1]) + abs(points[i][0] - points[j][0]);
                    minHeap.push({-dist, j});
                }
            }
            while(visited[minHeap.top().second])
                minHeap.pop();
            ans -= minHeap.top().first;
            i = minHeap.top().second;
            minHeap.pop();
        }
        return ans;
    }
};

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        i, ans, cp = 0, 0, 1 # cp- connected points
        visited = [False] * n
        minHeap = []
        
        while cp < n:
            visited[i] = True
            
            for j in range(n):
                if not visited[j]:
                    dist = abs(points[i][1] - points[j][1]) + abs(points[i][0] - points[j][0])
                    heappush(minHeap, (dist, j))
                
            
            while visited[minHeap[0][1]]:
                heappop(minHeap)
                
            ans += minHeap[0][0]
            i = minHeap[0][1]
            heappop(minHeap)
            
            cp += 1
            
        return ans
    